pkg/extract/redhat/csaf/csaf.go

```diff
@@ -680,216 +687,354 @@ func walkVulnerabilities(vulns []csaf.Vulnerability, pids []csaf.ProductID) (map
+	}
+}
+
+func buildVulnerabilities(baseVulnerabilities map[string]vulnerabilityContentTypes.Content, vassm map[string]map[csaf.ProductID]ass, pmmm map[string]map[ass]map[string]productsWithMaxPid) ([]vulnerabilityTypes.Vulnerability, error) {
+	// CVE ID -> severity -> major -> max Pid list
+	vpmmm := make(map[string]map[severity]map[string][]string)
+	for cveID, assm := range vassm {
+		for _, ass := range assm {
+			vpmm, found := vpmmm[cveID]
+			if !found {
+				vpmm = make(map[severity]map[string][]string)
+			}
+			vpm, found := vpmm[ass.severity]
+			if !found {
+				vpm = make(map[string][]string)
+			}
+
+			apm := func() map[string]productsWithMaxPid {
+				apmm, found := pmmm[cveID]
+				if !found {
+					return make(map[string]productsWithMaxPid)
 				}
-				if ass.severity.impact != "" {
-					ss = append(ss, severityTypes.Severity{
-						Type:   severityTypes.SeverityTypeVendor,
-						Source: "secalert@redhat.com",
-						Vendor: &ass.severity.impact,
-					})
+				apm, found := apmm[ass]
+				if !found {
+					return make(map[string]productsWithMaxPid)
 				}
-				return ss, nil
+				return apm
 			}()
+			for major, pmax := range apm {
+				if !slices.Contains(vpm[major], string(pmax.maxPid)) {
+					vpm[major] = append(vpm[major], string(pmax.maxPid))
+				}
+			}
+			vpmm[ass.severity] = vpm
+			vpmmm[cveID] = vpmm
+		}
+	}
+
+	var vs []vulnerabilityTypes.Vulnerability
+	for cveID, vpmm := range vpmmm {
+		baseV, found := baseVulnerabilities[cveID]
+		if !found {
+			return nil, errors.Errorf("no base vulnerability found for %s", cveID)
+		}
+
+		for sev, vpm := range vpmm {
+			sevs, err := buildSeverities(baseV, sev)
 			if err != nil {
-				return nil, nil, nil, errors.Wrap(err, "walk severity")
+				return nil, errors.Wrap(err, "build severities")
 			}
-			vuln.Severity = ss
+			baseV.Severity = sevs
 
-			vs = append(vs, vulnerabilityTypes.Vulnerability{
-				Content: vuln,
-				Segments: func() []segmentTypes.Segment {
-					var ss []segmentTypes.Segment
-					for v := range vpvm[string(vuln.ID)][pid] {
-						ss = append(ss, segmentTypes.Segment{
-							Ecosystem: ecosystemTypes.Ecosystem(fmt.Sprintf("%s:%s", ecosystemTypes.EcosystemTypeRedHat, v)),
-							Tag:       segmentTypes.DetectionTag(fmt.Sprintf("%s:%s", string(vuln.ID), pid)),
-						})
+			var ss []segmentTypes.Segment
+			for major, maxPids := range vpm {
+				for _, maxPid := range maxPids {
+					s := segmentTypes.Segment{
+						Ecosystem: ecosystemTypes.Ecosystem(fmt.Sprintf("%s:%s", ecosystemTypes.EcosystemTypeRedHat, major)),
+						Tag:       calculateTag(cveID, csaf.ProductID(maxPid)),
 					}
-					return ss
-				}(),
+					ss = append(ss, s)
+				}
+			}
+			vs = append(vs, vulnerabilityTypes.Vulnerability{
+				Content:  baseV,
+				Segments: ss,
 			})
+		}
+	}
+
+	return vs, nil
+}
 
+func buildSeverities(baseVulnerability vulnerabilityContentTypes.Content, sev severity) ([]severityTypes.Severity, error) {
```

Comment:
unused baseVulnerability
```suggestion
func buildSeverities(sev severity) ([]severityTypes.Severity, error) {
```